firebase-firestore/src/main/java/com/google/firebase/firestore/util/TaskQueue.java

```diff
@@ -0,0 +1,120 @@
+
+/**
+ * A queue for parallel execution of independent tasks.
+ *
+ * <p>All tasks must return the same result type T and results can be be fetched via
+ * `awaitResults()`. It is an error to enqueue tasks after invoking `awaitResults()`.
+ *
+ * <p>This class is not thread-safe. All public methods must be called from the same thread.
+ */
+public class TaskQueue<T> {
+  private BlockingQueue<TaskResult<T>> taskResults = new PriorityBlockingQueue<>();
+  private Semaphore completedTasks = new Semaphore(0);
+  private int totalTaskCount = 0;
+  private boolean acceptsTasks = true;
+
+  /** A single result from a background task. Encapsulates the task number to allow for sorting. */
+  private static class TaskResult<T> implements Comparable<TaskResult<T>> {
+    final int numTask;
+    final T result;
+    final Exception exception;
+
+    private TaskResult(int numTask, T result) {
+      this.numTask = numTask;
+      this.result = result;
+      this.exception = null;
+    }
+
+    private TaskResult(int numTask, Exception exception) {
+      this.numTask = numTask;
+      this.result = null;
+      this.exception = exception;
+    }
+
+    @Override
+    public int compareTo(TaskResult<T> other) {
+      return Integer.compare(numTask, other.numTask);
+    }
+  }
+
+  /** Enqueue a task on Android's THREAD_POOL_EXECUTOR. */
+  public void enqueueInBackground(Callable<T> task) {
+    hardAssert(acceptsTasks, "enqueueInBackground() called after awaitResults()");
+
+    int currentTask = ++totalTaskCount;
+    Executors.BACKGROUND_EXECUTOR.execute(
+        () -> {
+          try {
+            T result = task.call();
+            taskResults.add(new TaskResult<>(currentTask, result));
+          } catch (Exception e) {
+            taskResults.add(new TaskResult<>(currentTask, e));
+          }
+          completedTasks.release();
+        });
+  }
+
+  /** Executes a task inline, allowing for more efficient execution of small tasks. */
+  public void enqueueInline(Callable<T> callable) {
+    hardAssert(acceptsTasks, "enqueueInBackground() called after awaitResults()");
+
+    int currentTask = ++totalTaskCount;
+    try {
+      T result = callable.call();
+      taskResults.add(new TaskResult<>(currentTask, result));
+    } catch (Exception e) {
+      taskResults.add(new TaskResult<>(currentTask, e));
+    }
+    completedTasks.release();
+  }
+
+  /**
+   * Blocks on the execution of all tasks. Returns a list of results in the order that tasks where
+   * added.
+   */
+  public List<T> awaitResults() throws ExecutionException {
+    acceptsTasks = false;
+
+    List<T> allResults = new ArrayList<>(totalTaskCount);
+    try {
+      completedTasks.acquire(totalTaskCount);
+      while (!taskResults.isEmpty()) {
+        TaskResult<T> currentResult = taskResults.take();
+
+        if (currentResult.exception != null) {
+          throw new ExecutionException("Unhandled exception in task", currentResult.exception);
```

Comment:
Assuming task result ordering doesn't matter, since your policy here is just to throw on the first exception we encountered you could save all this translation code (and the TaskResult type declaration even) if you made the completedTasks buffer just contain `T` (and then stored the first exception separately).